### PR TITLE
SI-6673 fixes macro problems with eta expansions

### PR DESCRIPTION
**Pull request by xeno-by**
Eta expansions previously caused the typer to disable macros. That was
done in order to detect eta expansion of macro defs and show the user
an appropriate error message.

Macros were disabled because to find out whether we're expanding
a macro def, we need to get its symbol, and to get a symbol of something
we need to typecheck that something. However typechecking automatically
expands macros, so, unless we disable macros, after a typecheck we won't
be able to analyze macro occurrences anymore.

Unfortunately this solution has a fatal flaw. By disabling macros we
not only prevent the eta-expandee from macro expanding, but also all
the subtrees of that eta-expandee (see SI-6673).

This commit adds a mechanism for fine-grained control over macro
expansion. Now it's possible to prohibit only the node, but not its
children from macro expanding.

### CODE REVIEW

**Comment by xeno-by** [comment]
review @retronym

**Comment by xeno-by** [comment]
This is a regression since 2.9.2, so I thought the fix would be appropriate for inclusion in 2.10.0-RC3.

**Comment by retronym** [comment]
patch itself lgtm, @adriaanm should decide on 2.10.0/1

File: src/reflect/scala/reflect/internal/StdAttachments.scala [hunk 1]
```diff
@@ -24,4 +24,6 @@ trait StdAttachments {
   case class CompoundTypeTreeOriginalAttachment(parents: List[Tree], stats: List[Tree])
 
   case class MacroExpansionAttachment(original: Tree)
+
+  case object SuppressMacroExpansionAttachment
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/StdAttachments.scala:28]: A little scaladoc would be warranted for these attachments.

**Comment by adriaanm** [comment]
This should be rare and avoidable enough to wait for 2.10.1. Also, I agree with @retronym on attaching some scaladocs to the attachments.

**Comment by xeno-by** [comment]
https://github.com/scala/scala/pull/1645

**Pull request closed without merging**